mrcnn/model.py

```diff
@@ -1933,9 +1935,8 @@ def data_generator(self,image_ids):
                         batch_mrcnn_class_ids, -1)
                     outputs.extend(
                         [batch_mrcnn_class_ids, batch_mrcnn_bbox, batch_mrcnn_mask])
-            if self.config.NUM_IMAGE_CLASSES:
-                inputs.append(batch_gt_image_classes)
-                outputs.append(batch_gt_image_classes)
+                    if self.config.NUM_IMAGE_CLASSES:
+                        outputs.append(batch_gt_image_classes)
```

Comment:
> So, I had another look through the Code, as per [doc](https://github.com/bertsky/Mask_RCNN/blob/8f9193807c2e77d70a0ec623a367d80aa30c6eac/mrcnn/model.py#L1776-L1778) during training `outputs` should be an empty array.

Ah, yes, of course! I completely forgot about that. This implementation uses the `Model.add_loss` API exclusively, i.e. it passes `None` to the `loss` argument of `compile` for every output:
https://github.com/matterport/Mask_RCNN/blob/3deaec5d902d16e1daf56b62d5971d428dc920bc/mrcnn/model.py#L2188

So, indeed, in the normal training regime, we must not pass any outputs (because no loss must be computed from these outputs).

> The `detection_targets` is a debug variable to get some debug outputs, so our new output should only be there if the debug output is there.
> 
> And as per lines [L1842-L1844](https://github.com/bertsky/Mask_RCNN/blob/8f9193807c2e77d70a0ec623a367d80aa30c6eac/mrcnn/model.py#L1842-L1844) to generate the detection targets one needs the `rpn_rois` which only gets generated if `random_rois > 0`.
> 
> That said, we should consider whether it is at all necessary to add the `image_class` to the output at all.

I still think we should add the **tensor variable** to the output IMO. But in the generator, we need to keep the indentation (under the scope of `random_rois` and `detection_targets`), so it stays away from the output tuple.